flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.runtime.io.network.buffer.BufferPool;
+import org.apache.flink.runtime.io.network.buffer.LocalBufferPool;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.TierMemorySpec;
+import org.apache.flink.util.ExceptionUtils;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * The implementation for {@link TieredStorageMemoryManager}. This is to request or recycle buffers
+ * from {@link LocalBufferPool} for different tiers or {@link BufferAccumulator}.
+ *
+ * <p>The buffers in a buffer pool are mainly divided into types, the exclusive buffers for a tier
+ * and the shared buffers for all tiers. Each tier decides the number of exclusive buffers and
+ * whether it can use the shared buffers by {@link TierMemorySpec}.
+ *
+ * <p>The buffers of {@link BufferAccumulator} is also managed in the {@link
+ * TieredStorageMemoryManager}. When {@link BufferAccumulator} requests or recycles buffers, it uses
+ * separate methods that are different from tiers.
+ *
+ * <p>Note that when initializing the {@link TieredStorageMemoryManagerImpl}, the {@link
+ * TierMemorySpec} of each tier should be ready. These tier memory specs mainly decide how many
+ * buffers can be used for every tier.
+ */
+public class TieredStorageMemoryManagerImpl implements TieredStorageMemoryManager {
```

Comment:
How does this memory manager guarantees the exclusive buffers? I see this class bookkeeps the number of buffers each tier and accumulator uses. However, I don't see any logic for reversing exclusive buffers for a tier, or limiting a tier from requesting too many buffers.

Reply:
I have updated the code according to the offline discussion. The main method of available buffers of a owner is `numAvailableBuffers`, which mainly to limiting a tier (`isMemoryReleasable` of the tier is `false`) from requesting too many buffers.